### PR TITLE
[block] add support for notification suppression

### PR DESCRIPTION
**Pull request by serban300**
# Reason for This PR

#2635 

### Performance Results

**Test setup:** We run on a m5d.metal host with Ubuntu and kernel 5.8.18. We start a number of microVMs with 8vCPUs and 16GB of RAM running linux kernel v5.8.18 as a guest OS. Each microVMs is pinned to NUMA node 1. We mount a ramdisk and we pin it to NUMA node 1 as well. For each microVM we create an ext4 file on the ramdisk and we attach it to the microVM as /dev/vdb. We want to have the content of the drives in the page cache in order to asses the performance improvement brought by notification suppression. Otherwise we would be limited by the disk access speed. Inside the VM we run the following command:

```
fio --name=test --ioengine=libaio --rw=readwrite --randrepeat=1 --bs=4k --size=500m --numjobs=1 --iodepth=32 --runtime=30 --time_base=1 --end_fsync=1 --direct=1 --filename=/dev/vdb
```

Current implementation:
|description|IOPS|VMM thread CPU load|kworker threads CPU load|IOPS per CPU load|
|-|-|-|-|-|
|1 VMs|104.86|0.36|0.13|213.5|
|4 VMs|439.28|1.51|0.52|216.34|
|8 VMs|849.83|3.1|1.07|203.47|
|16 VMs|1524.45|6.76|2.24|169.51|
|32 VMs|2202.07|8.32|2.19|209.51|
|64 VMs|1182.12|5.76|1.77|156.89|

Notification suppression:
|description|IOPS|VMM thread CPU load|kworker threads CPU load|IOPS per CPU load|increase in efficiency|
|-|-|-|-|-|-|
|1 VMs|160.72|0.42|0.09|315.65|47.85%|
|4 VMs|599.36|1.64|0.36|299.79|38.57%|
|8 VMs|1175.92|3.33|0.81|283.79|39.48%|
|16 VMs|1872.87|7.07|2.12|203.68|20.16%|
|32 VMs|2551.24|8.67|2.61|226.2|7.97%|
|64 VMs|1369.84|6.07|1.79|174.26|11.07%|

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [x] All commits in this PR are signed (`git commit -s`).
- [x] The reason for this PR is clearly provided (issue no. or explanation).
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] Any newly added `unsafe` code is properly documented.
- [x] Any API changes are reflected in `firecracker/swagger.yaml`.
- [x] Any user-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.

### CODE REVIEW

**Comment by serban300** [comment]
I made some small adjustments to the notification suppression functionality and also implemented the logic that breaks snapshot compatibility with previous versions of Firecracker. The PR is now ready for review. @acatangiu @sandreim @firecracker-microvm/compute-capsule PTAL when you have time !

File: src/devices/src/virtio/queue.rs [hunk 1]
```diff
@@ -296,6 +303,32 @@ impl Queue {
             return None;
         }
 
+        self.do_pop(mem)
+    }
+
+    /// Try to pop the first available descriptor chain from the avail ring.
+    /// If no descriptor is available, enable notifications.
+    pub fn pop_or_enable_notification<'a, 'b>(
```
Comment (sandreim) [inline src/devices/src/virtio/queue.rs:311]: Do we want to have a future proof approach here so that we end up in an ideal state of being able to quickly switch to rust-vmm implementation ? https://github.com/rust-vmm/vm-virtio/blob/master/crates/virtio-queue/src/lib.rs

**Review by sandreim — COMMENTED** [review]
These changes are properly structured and look good. I've left some comments. I have only one concern related to diverging from the current rust-vmm state of virtio queue implementation. We should try to be in sync with the upstream design so that it easier to switch. For example I think the Iterators semantics are more appropriate and easier to read or get wrong than queue.pop(). 

What's your thinking around the effort of sync this PR to rust-vmm ?

File: src/devices/src/virtio/queue.rs (hunk 1 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:311]: Yes, of course. Actually this implementation is my interpretation of an idea suggested by @alexandruag who works closely with the rust-vmm community. I'm not sure how easy it would be to implement this in an iterator, but this way of enabling notifications simplifies the devices implementation by a lot. Otherwise in the devices we would have something like:

```
outter': loop {
    while let Some(chain) = queue.pop() {
        if rate_limit(chain) {
            break outter;
        }

        process(chain);
    }

    if !queue.enable_notification() {
        break;
    }
}
``` 

Especially in the net device this kind of implementation would be really hard to follow and I think it would lead to complications.

File: src/devices/src/virtio/queue.rs [hunk 2]
```diff
@@ -393,6 +401,64 @@ impl Queue {
...
+    /// Fetch the used event (`virtq_avail->used_event`) from guest memory.
+    /// This field is used when notification suppression is enabled. It is written by the driver,
+    /// in order to specify how far the device can progress before a notification is required.
+    #[inline(always)]
+    pub fn used_event(&self, mem: &GuestMemoryMmap) -> Wrapping<u16> {
+        // We need to find the `used_event` field from the avail ring.
+        let used_event_addr = 4 + 2 * self.actual_size();
+
+        // Complete all the writes in add_used() before reading the event.
+        fence(Ordering::SeqCst);
```
[the remark this answers is not in the document]
Comment (serban300) [reply src/devices/src/virtio/queue.rs]: AFAIK `Ordering::Acquire` is not enough. Here we need a store-load barrier (we need to make sure that the store to `used_idx` was completed before the following load. So we need to actually stop the current CPU until all the stores are visible to all other CPUs. This can be done only with a SeqCst barrier as far as i know. `Ordering::Acquire` only provides a load-load barrier when coupled with a  `Ordering::Release` fence.

File: src/devices/src/virtio/queue.rs (hunk 1 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:311]: As we discussed offline, this shouldn't break compatibility with rust-vmm.

File: tests/integration_tests/functional/test_snapshot_version.py [hunk 3]
```diff
@@ -186,3 +153,48 @@ def test_create_invalid_version(bin_cloner_path):
...
+    _snapshot_dir = snapshot_builder.create_snapshot_dir()
+
+    # Pause microVM for snapshot.
+    response = test_microvm.vm.patch(state='Paused')
+    assert test_microvm.api_session.is_status_no_content(response.status_code)
+
+    # We try to create a snapshot to a target version < 0.26.0.
+    # This should fail because Fc versions < 0.26.0 don't support
+    # virtio notification suppression.
+    response = test_microvm.snapshot.create(
```
Comment (sandreim) [inline tests/integration_tests/functional/test_snapshot_version.py:185]: We could test v0.23 - v0.25 here to have full coverage.

File: src/devices/src/virtio/queue.rs [hunk 4]
```diff
@@ -393,6 +427,82 @@ impl Queue {
...
+
+    /// Try to enable notification events from the guest driver. Returns true if notifications were
+    /// successfully enabled. Otherwise it means that one or more descriptors can still be consumed
+    /// from the available ring and we can't guarantee that there will be a notification. In this
+    /// case the caller might want to consume the mentioned descriptors and call this method again.
+    pub fn try_enable_notification(&mut self, mem: &GuestMemoryMmap) -> bool {
+        // If the device doesn't use notification suppression, we'll continue to get notifications
+        // no matter what.
+        if !self.uses_notif_suppression {
+            return true;
```
Comment (sandreim) [inline src/devices/src/virtio/queue.rs:459]: Why are we double checking this ? In pop_or_enable_notification() we are already doing this.

File: src/devices/src/virtio/queue.rs [hunk 5]
```diff
@@ -393,6 +427,82 @@ impl Queue {
...
+        // Make sure all subsequent reads are performed after `set_avail_event`.
+        fence(Ordering::SeqCst);
+
+        // If the actual avail_idx is different than next_avail one or more descriptors can still
+        // be consumed from the available ring.
+        self.next_avail.0 == self.avail_idx(mem).0
+    }
+
+    /// Enable/disable notification suppression.
+    pub fn toggle_notif_suppression(&mut self, enabled: bool) {
```
Comment (sandreim) [inline src/devices/src/virtio/queue.rs:478]: Is there any usecase where we would want to disable ?

File: src/devices/src/virtio/queue.rs [hunk 6]
```diff
@@ -393,6 +427,82 @@ impl Queue {
...
+    }
+
+    /// Check if we need to kick the guest.
+    ///
+    /// Please note this method has side effects: once it returns `true`, it considers the
+    /// driver will actually be notified, and won't return `true` again until the driver
+    /// updates `used_event` and/or the notification conditions hold once more.
+    ///
+    /// This is similar to the `vring_need_event()` method implemented by the Linux kernel.
+    pub fn prepare_kick(&mut self, mem: &GuestMemoryMmap) -> bool {
```
Comment (sandreim) [inline src/devices/src/virtio/queue.rs:489]: Nit: should we rename to `needs_kick` ?

File: src/devices/src/virtio/queue.rs (hunk 4 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:459]: Both methods are public so they can be used independently, and they need to treat this case differently.

File: src/devices/src/virtio/queue.rs (hunk 5 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:478]: No. I changed the method to `enable_notification()`.

File: src/devices/src/virtio/queue.rs (hunk 6 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:489]: IMO `needs_kick` might imply that we are performing only a check, but we are actually also changing the state here. That's why I named it `prepare_kick`. But if you have strong preferences I can rename it.

File: src/vmm/src/persist.rs [hunk 7]
```diff
@@ -297,23 +306,39 @@ fn snapshot_memory_to_file(
...
+
+    if data_version < FC_V0_26_SNAP_VERSION {
+        vmm.mmio_device_manager
+            .for_each_virtio_device(|_id, _info, dev| {
+                if dev
+                    .lock()
+                    .expect("Poisoned lock")
+                    .has_feature(u64::from(VIRTIO_RING_F_EVENT_IDX))
+                {
+                    return Err(CreateSnapshotError::IncompatibleDeviceFeatures);
```
[the remark this answers is not in the document]
Comment (serban300) [reply src/vmm/src/persist.rs:335]: Done. I added a string to it in order to mention the feature that is causing the problem.

File: tests/integration_tests/functional/test_snapshot_version.py (hunk 3 above)
Comment (serban300) [reply tests/integration_tests/functional/test_snapshot_version.py:185]: Done.

**Review by sandreim — DISMISSED** [review]

File: src/devices/src/virtio/queue.rs [hunk 8]
```diff
@@ -393,6 +401,82 @@ impl Queue {
...
+
+        // Make sure all subsequent reads are performed after `set_avail_event`.
+        fence(Ordering::SeqCst);
+
+        // If the actual avail_idx is different than next_avail one or more descriptors can still
+        // be consumed from the available ring.
+        self.next_avail.0 == self.avail_idx(mem).0
+    }
+
+    /// Enable/disable notification suppression.
```
Comment (acatangiu) [inline src/devices/src/virtio/queue.rs:477]: nit: incorrect doc: this is not a toggle, only enables never disables

File: src/devices/src/virtio/queue.rs [hunk 9]
```diff
@@ -393,6 +401,82 @@ impl Queue {
...
+        }
+
+        // We need to expose used array entries before checking the used_event.
+        fence(Ordering::SeqCst);
+
+        let new = self.next_used;
+        let old = self.next_used - self.num_added;
+        let used_event = self.used_event(mem);
+
+        self.num_added = Wrapping(0);
```
Comment (acatangiu) [inline src/devices/src/virtio/queue.rs:502]: just an idea: since this is the only state change (side effect) of this function, would it make sense to change this fn to `needs_kick()` which only does checking (with no side effects) and do `self.num_added = 0` in actual `do_kick()` logic?

File: src/devices/src/virtio/queue.rs [hunk 10]
```diff
@@ -303,6 +303,32 @@ impl Queue {
...
+
+        self.do_pop(mem)
+    }
+
+    /// Pop the first available descriptor chain from the avail ring.
+    ///
+    /// # Important
+    /// This is an internal method that ASSUMES THAT THERE ARE AVAILABLE DESCRIPTORS. Otherwise it
+    /// will retrieve a descriptor that contains garbage data (obsolete/empty).
+    fn do_pop<'a, 'b>(&'a mut self, mem: &'b GuestMemoryMmap) -> Option<DescriptorChain<'b>> {
```
Comment (acatangiu) [inline src/devices/src/virtio/queue.rs:331]: nit: rename `do_pop_unchecked()`

File: src/devices/src/virtio/block/device.rs [hunk 11]
```diff
@@ -320,16 +321,19 @@ impl Block {
...
                 )
             });
+
+            if queue.prepare_kick(mem) {
+                self.irq_trigger
+                    .trigger_irq(IrqType::Vring)
+                    .unwrap_or_else(|_| {
+                        METRICS.block.event_fails.inc();
+                    });
+            }
```
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:331]: This introduces a regression for cases where notif suppression is not enabled/negotiated: with this new code, IRQs are being generated _on each descriptor_ instead of just one IRQ after the full chunk of currently avail descriptors.

I am expecting this regression to be actually visible in perf testing with notif suppres disabled/unavailable in the guest.

File: src/vmm/src/device_manager/mmio.rs [hunk 12]
```diff
@@ -360,6 +361,32 @@ impl MMIODeviceManager {
         Ok(())
     }
 
+    /// Run fn for each registered device.
```
Comment (acatangiu) [inline src/vmm/src/device_manager/mmio.rs:364]:
```suggestion
    /// Run fn for each registered virtio device.
```

File: src/vmm/src/device_manager/mmio.rs [hunk 13]
```diff
@@ -360,6 +361,32 @@ impl MMIODeviceManager {
...
+    pub fn for_each_virtio_device<F, E>(&self, mut f: F) -> std::result::Result<(), E>
```
Comment (acatangiu) [inline src/vmm/src/device_manager/mmio.rs:365]: You can use this newly added specialized fn to simplify code in `fn kick_devices()` below and possibly even `fn with_virtio_device_with_id()`

File: src/vmm/src/persist.rs [hunk 14]
```diff
@@ -32,10 +32,12 @@ use seccompiler::BpfThreadMap;
 use snapshot::Snapshot;
 use versionize::{VersionMap, Versionize, VersionizeResult};
 use versionize_derive::Versionize;
+use virtio_gen::virtio_ring::VIRTIO_RING_F_EVENT_IDX;
 use vm_memory::GuestMemoryMmap;
 
 #[cfg(target_arch = "x86_64")]
 const FC_V0_23_SNAP_VERSION: u16 = 1;
+const FC_V0_26_SNAP_VERSION: u16 = 4;
```
Comment (acatangiu) [inline src/vmm/src/persist.rs:40]: how does the version encoding numbering work? just looking at this code here it looks like we're jumping from version `1` to version `4`.. What am I missing, and how can we make this clearer for future PRs?

(this comment can also be addressed in a separate PR with its own issue)

File: src/vmm/src/version_map.rs [hunk 15]
```diff
@@ -41,6 +45,7 @@ lazy_static! {
 
         mapping.insert(String::from("0.24.0"), 2);
         mapping.insert(String::from("0.25.0"), 3);
+        mapping.insert(String::from("0.26.0"), 4);
```
Comment (acatangiu) [inline src/vmm/src/version_map.rs:48]: I see here the full mapping. Since we're using also constants (`FC_V0_23_SNAP_VERSION`), we should centralize their definitions and also use them here.

**Review by acatangiu — COMMENTED** [review]
Great stuff!

File: src/devices/src/virtio/block/device.rs (hunk 11 above)
Comment (acatangiu) [reply src/devices/src/virtio/block/device.rs:331]: I believe we can structure the code so as to avoid regressions on the _legacy_ path.

File: src/devices/src/virtio/queue.rs (hunk 8 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:477]: Fixed.

File: src/devices/src/virtio/queue.rs (hunk 9 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:502]: This is better from a naming point of view, but it introduces the possibility of forgetting to call `do_kick()` after `needds_kick()`. Actually the code was similar to this in an earlier version of the PR and when I tried to use this logic to implement notif suppression on net also in order to do some experiments I forgot to call `do_kick()` and it lead to freezes. I think it's safer to keep it as it is now.

File: src/vmm/src/version_map.rs (hunk 15 above)
Comment (serban300) [reply src/vmm/src/version_map.rs:48]: Nice. Fixed.

File: src/devices/src/virtio/block/device.rs (hunk 11 above)
Comment (serban300) [reply src/devices/src/virtio/block/device.rs:331]: I was unsure if we should treat this case. It complicates the code a bit, and notification suppression is present in all the kernels that we support AFAIK. Anyway I tried to fix this. I added a new commit. PTAL.

File: src/devices/src/virtio/queue.rs (hunk 10 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:331]: Nice. Done.

File: src/vmm/src/device_manager/mmio.rs (hunk 12 above)
Comment (serban300) [reply src/vmm/src/device_manager/mmio.rs:364]: Fixed.

File: src/vmm/src/device_manager/mmio.rs (hunk 13 above)
Comment (serban300) [reply src/vmm/src/device_manager/mmio.rs:365]: Nice. I used it in `kick_devices()`. `with_virtio_device_with_id` uses `get_device` which is more efficient.

File: src/vmm/src/persist.rs (hunk 14 above)
Comment (serban300) [reply src/vmm/src/persist.rs:40]: Fixed.

File: src/devices/src/virtio/queue.rs (hunk 9 above)
Comment (acatangiu) [reply src/devices/src/virtio/queue.rs:509]: I still think current code is not ideal since it does only half of something. My preference was that we either:
- have single function which checks if IRQ needed then also kicks
- cleanly split checking and kicking

But I see that it's not easy to do that since part of the logic is in the queue while irq logic is in the device. I will not block the PR on this since it's a two-way door implementation detail.

File: src/vmm/src/persist.rs [hunk 16]
```diff
@@ -32,10 +34,9 @@ use seccompiler::BpfThreadMap;
 use snapshot::Snapshot;
 use versionize::{VersionMap, Versionize, VersionizeResult};
 use versionize_derive::Versionize;
+use virtio_gen::virtio_ring::VIRTIO_RING_F_EVENT_IDX;
 use vm_memory::GuestMemoryMmap;
 
-#[cfg(target_arch = "x86_64")]
-const FC_V0_23_SNAP_VERSION: u16 = 1;
 #[cfg(target_arch = "x86_64")]
 const FC_V0_23_MAX_DEVICES: u32 = 11;
```
Comment (acatangiu) [inline src/vmm/src/persist.rs:41]: nit: should this also go in `version_map.rs` const definitions?

File: src/devices/src/virtio/block/device.rs [hunk 17]
```diff
@@ -323,16 +359,14 @@ impl Block {
             });
 
             if queue.prepare_kick(mem) {
-                self.irq_trigger
-                    .trigger_irq(IrqType::Vring)
-                    .unwrap_or_else(|_| {
-                        METRICS.block.event_fails.inc();
-                    });
+                self.irq_trigger.trigger_vring_irq(batch_irqs);
```
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:362]: To keep the code simple, could we "delay" the IRQ here for both interrupt suppression and no-suppression cases?

Basically "batch" IRQs in both cases? (might introduce extra latency for suppression case, but not sure if that's all or if it messes with the mechanism in other ways)

File: src/devices/src/virtio/block/device.rs [hunk 18]
```diff
@@ -177,6 +178,41 @@ impl Drop for DiskProperties {
     }
 }
 
+pub(crate) struct BlockIrqTrigger {
+    pub(crate) inner: IrqTrigger,
+    has_batched_vring_irqs: bool,
```
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:183]:
```suggestion
    pending_batched_vring_irqs: bool,
```

File: src/devices/src/virtio/block/device.rs [hunk 19]
```diff
@@ -177,6 +178,41 @@ impl Drop for DiskProperties {
...
+        } else {
+            self.do_trigger_vring_irq();
+        }
+    }
+
+    fn flush_vring_irqs(&mut self) {
+        if self.has_batched_vring_irqs {
+            self.do_trigger_vring_irq();
+        }
+    }
```
Comment (acatangiu) [inline src/devices/src/virtio/block/device.rs:213]: For better clarity, I would try something like:
```suggestion
    pub fn trigger_vring_irq(&mut self) {
        self.inner.trigger_irq(IrqType::Vring).unwrap_or_else(|_| {
            METRICS.block.event_fails.inc();
        });
    }

    pub fn enqueue_vring_irq(&mut self) {
        self.has_batched_vring_irqs = true;
    }

    pub fn flush_vring_irqs(&mut self) {
        if self.has_batched_vring_irqs {
            self.trigger_vring_irq();
            self.has_batched_vring_irqs = false;
        }
    }
}
```

Then call `enqueue_vring_irq()` or directly `trigger_vring_irq()` in block based on batching flag, instead of passing down batching flag to this object.

File: src/devices/src/virtio/block/device.rs (hunk 17 above)
Comment (acatangiu) [reply src/devices/src/virtio/block/device.rs:362]: If we can't delay, then I suggest we call `enqueue_irq()` or `trigger_irq()` here based on `batch_irqs` flag to keep logic clear and centered:

```suggestion
                if batch_irqs {
                    self.irq_trigger.enqueue_vring_irq();
                } else {
                    self.irq_trigger.trigger_vring_irq();
                }
```

File: src/devices/src/virtio/block/device.rs (hunk 18 above)
Comment (serban300) [reply src/devices/src/virtio/block/device.rs:183]: Removed this code as discussed offline.

File: src/devices/src/virtio/block/device.rs (hunk 19 above)
Comment (serban300) [reply src/devices/src/virtio/block/device.rs:213]: Removed this code as discussed offline.

File: src/vmm/src/persist.rs (hunk 16 above)
Comment (serban300) [reply src/vmm/src/persist.rs:41]: I'm not sure. Since the check is implemented in this file I would keep it here. But if you have strong preferences I can move it.

File: src/devices/src/virtio/block/device.rs (hunk 17 above)
Comment (serban300) [reply src/devices/src/virtio/block/device.rs:362]: As we discussed offline we decided to drop batching for both cases (with and wo notif suppression). This will help us keep the code simple.

Since all kernels >=4.14 support notification suppression, there are only few corner cases where the performance regression would show:
- very old or custom kernel that doesn't support notif suppression
- new kernel with notif suppression disabled (for snapshot compatibility for example)

So for the moment we'll drop batching and revisit it later if needed.

File: src/devices/src/virtio/queue.rs (hunk 9 above)
Comment (serban300) [reply src/devices/src/virtio/queue.rs:509]: Yes, exactly. This is the first half of a kick as specified [here](https://elixir.bootlin.com/linux/v5.13/source/drivers/virtio/virtio_ring.c#L1814). It's also used in the same way in the [kernel block device](https://elixir.bootlin.com/linux/v5.13/source/drivers/block/virtio_blk.c#L210) btw. 

Another option is to have a method `Queue::kick(&self, irq_trigger: IrqTrigger)`. This way it could check if a notification is needed and call `irq_trigger.trigger_irq()` if so. But it would probably raise question related to rust-vmm upstream compatibility. Also the IrqTrigger is not mature enough and I'm not sure how it will evolve and if it will be so easy to pass it to the `Queue::kick()` in the future.

Persoanlly for the moment I think it would be safest to keep it as it is and maybe revisit this after migrating to the Queue implementation in rust-vmm.

**Review by acatangiu — APPROVED** [review]
🎆

**Review by sandreim — APPROVED** [review]
🚀
